workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToCreatedReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToEnable = new HashSet<MonoBehaviour>();
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToDisable = new HashSet<MonoBehaviour>();
 
-        public SpatialOSBehaviourManager(GameObject gameObject)
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> componentIdToCreatedReadersWriters =
+            new Dictionary<uint, HashSet<IReaderInternal>>();
+
+        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+
+        public SpatialOSBehaviourManager(GameObject gameObject, SpatialOSBehaviourLibrary library)
         {
-            // Setup caches
+            behaviourLibrary = library;
+
+            var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
+            entity = spatialComponent.Entity;
+            spatialId = spatialComponent.SpatialEntityId;
+
+            foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
+            {
+                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
+                foreach (var id in readerIds)
+                {
+                    if (!componentIdToReaderSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToReaderSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToReaderSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                var writerIds = library.GetRequiredWriterComponentIds(behaviour.GetType());
+                foreach (var id in writerIds)
+                {
+                    if (!componentIdToWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToWriterSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                numUnsatisfiedReadersOrWriters[behaviour] = readerIds.Count + writerIds.Count;
+            }
         }
 
-        public List<ReaderWriterImpl> GetReadersWriters(uint componentId)
+
+        public HashSet<IReaderInternal> GetReadersWriters(uint componentId)
```

Comment:
What does this function add over just doing the lookup directly?

Reply:
This is just to have a public interface rather than being tied to a particular way of storing internally